### PR TITLE
Add support for custom weave data folder

### PR DESCRIPTION
**Pull request by ldmberman**
The motivation is to be able to run multiple independent miners from the same folder for testing purposes. As a bonus, the feature simplifies the task of copying data between machines (a common miners need).

### CODE REVIEW

File: src/ar_storage.erl [hunk 1]
```diff
@@ -29,16 +29,17 @@ start() ->
...
-	filelib:ensure_dir(?BLOCK_DIR),
-	filelib:ensure_dir(?BLOCK_ENC_DIR),
-	filelib:ensure_dir(?WALLET_LIST_DIR),
-	filelib:ensure_dir(?HASH_LIST_DIR),
+	DataDir = ar_meta_db:get(data_dir) ++ "/",
+	filelib:ensure_dir(DataDir ++ ?TX_DIR ++ "/"),
+	filelib:ensure_dir(DataDir ++ ?BLOCK_DIR ++ "/"),
+	filelib:ensure_dir(DataDir ++ ?BLOCK_ENC_DIR ++ "/"),
+	filelib:ensure_dir(DataDir ++ ?WALLET_LIST_DIR ++ "/"),
+	filelib:ensure_dir(DataDir ++ ?HASH_LIST_DIR ++ "/"),
```
Comment (martin-torhage) [inline src/ar_storage.erl]: Shouldn't also the WALLET_DIR be created here?

File: arweave-server [hunk 2]
```diff
@@ -2,7 +2,7 @@
 
 set -o errexit
 
-mkdir -p ebin logs blocks hash_lists wallets wallet_lists txs blocks/enc priv data data/mnesia
+mkdir -p ebin logs priv
```
Comment (martin-torhage) [inline arweave-server]: Why is this needed? All three exists in git.

**Comment by martin-torhage** [comment]
Some overall feedback/questions:

- Did you consider using `filename:join/1`? I think it makes sense but maybe not so important.
- It seems like we're using `.gitignore` files to ensure directories exists. That's bad cargo-culting in my opinion. Let's do something explicit like using empty `.keep` files.

File: docker-arweave-server [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
 set -o errexit
 
-mkdir -p ebin logs blocks wallets txs blocks/enc priv data/genesis_txs
+mkdir -p ebin logs priv
```
Comment (martin-torhage) [inline docker-arweave-server]: Why is this needed? All three exists in git.

File: src/ar.erl [hunk 4]
```diff
@@ -227,6 +232,9 @@ start(
 	error_logger:logfile({open, Filename = generate_logfile_name()}),
 	error_logger:tty(false),
 
+	ar_meta_db:start(),
+	ar_meta_db:put(data_dir, DataDir),
```
Comment (martin-torhage) [inline src/ar.erl]: I would prefer if all `ar_meta_db:put/2` calls are moved up here so they stay together.

Btw, our style guide doesn't say anything about blank lines but I would like to not have any.

File: src/ar_storage.erl [hunk 5]
```diff
@@ -59,19 +60,20 @@ block_exists(Hash) ->
 %% @doc Move a block into the 'invalid' block directory.
 invalidate_block(B) ->
 	ar_block_index:remove(B#block.indep_hash),
+	DataDir = ar_meta_db:get(data_dir) ++ "/",
 	TargetFile =
 		lists:flatten(
 			io_lib:format(
 				"~s/invalid/~w_~s.json",
-				[?BLOCK_DIR, B#block.height, ar_util:encode(B#block.indep_hash)]
+				[DataDir ++ ?BLOCK_DIR, B#block.height, ar_util:encode(B#block.indep_hash)]
```
Comment (martin-torhage) [inline src/ar_storage.erl]: Let's put the whole `lists:flatten/1` call in a function that takes a block as argument.

File: src/ar_storage.erl [hunk 6]
```diff
@@ -325,7 +327,7 @@ write_tx(Tx) ->
 		Name = lists:flatten(
 			io_lib:format(
 				"~s/~s.json",
-				[?TX_DIR, ar_util:encode(Tx#tx.id)]
+				[ar_meta_db:get(data_dir) ++ "/" ++ ?TX_DIR, ar_util:encode(Tx#tx.id)]
```
Comment (martin-torhage) [inline src/ar_storage.erl]: I would like to put the whole `lists:flatten/1` call in a function that takes a transaction as argument.

File: src/ar_storage.erl [hunk 7]
```diff
@@ -29,16 +29,17 @@ start() ->
...
 ensure_directories() ->
-	filelib:ensure_dir(?TX_DIR),
-	filelib:ensure_dir(?BLOCK_DIR),
-	filelib:ensure_dir(?BLOCK_ENC_DIR),
-	filelib:ensure_dir(?WALLET_LIST_DIR),
-	filelib:ensure_dir(?HASH_LIST_DIR),
+	DataDir = ar_meta_db:get(data_dir) ++ "/",
+	filelib:ensure_dir(DataDir ++ ?TX_DIR ++ "/"),
+	filelib:ensure_dir(DataDir ++ ?BLOCK_DIR ++ "/"),
+	filelib:ensure_dir(DataDir ++ ?BLOCK_ENC_DIR ++ "/"),
```
Comment (martin-torhage) [inline src/ar_storage.erl]: I don't understand what "block enc" means.

File: src/ar_storage.erl [hunk 8]
```diff
@@ -29,16 +29,17 @@ start() ->
...
+	filelib:ensure_dir(DataDir ++ ?WALLET_LIST_DIR ++ "/"),
```
Comment (martin-torhage) [inline src/ar_storage.erl]: The slash in the end should not be needed. We have it in many places.

**Comment by martin-torhage** [comment]
I've added some inline comments. Here are the overall feedback:

 - Have you seen the `filename: join/1` function? It might be cleaner it many cases.
 - I suggest adding something like this in `ar` and use it as much as possible:
```
data_filename(Components) ->
    filename:join([ar_meta_db:get(data_dir) | Components]).
```

I lost one comment before I posted it. Will add it when I remember what it was about :)

**Comment by ldmberman** [comment]
Yes, `join` makes sense off course. I mostly preserved the style of the code as it used to be but I am happy to improve it.

File: src/ar_storage.erl [hunk 9]
```diff
@@ -29,16 +29,17 @@ start() ->
...
+	filelib:ensure_dir(DataDir ++ ?HASH_LIST_DIR ++ "/"),
 	filelib:ensure_dir(?LOG_DIR).
```
Comment (martin-torhage) [inline src/ar_storage.erl]: This dir already exists in git. Can we remove this line?

File: src/ar_storage.erl (hunk 1 above)
Comment (ldmberman) [reply src/ar_storage.erl]: `ar_wallet:new_keyfile` takes care of it.

File: docker-arweave-server (hunk 3 above)
Comment (ldmberman) [reply docker-arweave-server]: The Dockerfile which was supposed to be used with this script has a dedicated stage for running the server where it just copies some of the build artifacts from the other stage, without copying the whole git folder. Creating `ebin` and `priv` (unlike `logs`) looks unnecessary anyway but I would leave these scripts as-is until we specifically dedicate some time to update or remove them.

File: src/ar.erl (hunk 4 above)
Comment (ldmberman) [reply src/ar.erl]: I like the visual separation of the blocks, but comments can often be used for that. I am ok with not using blank lines.

File: src/ar_storage.erl (hunk 8 above)
Comment (ldmberman) [reply src/ar_storage.erl]: It is in fact needed here, to make `filelib:ensure_dir` create a directory if one does not exist.

File: src/ar.erl (hunk 4 above)
Comment (samcamwilliams) [reply src/ar.erl]: +1 for comments

File: src/ar_storage.erl (hunk 7 above)
Comment (ldmberman) [reply src/ar_storage.erl]: Renamed into `ENCRYPTED_BLOCK_DIR`, hope it's clear now.

File: src/ar_storage.erl (hunk 1 above)
Comment (martin-torhage) [reply src/ar_storage.erl]: ✅

File: docker-arweave-server (hunk 3 above)
Comment (martin-torhage) [reply docker-arweave-server]: ✅

File: arweave-server (hunk 2 above)
Comment (martin-torhage) [reply arweave-server]: ✅

File: src/ar_storage.erl (hunk 7 above)
Comment (martin-torhage) [reply src/ar_storage.erl]: ✅

File: src/ar_storage.erl (hunk 8 above)
Comment (martin-torhage) [reply src/ar_storage.erl]: Ohhh I made false assumptions ✅

File: src/ar.erl (hunk 4 above)
Comment (martin-torhage) [reply src/ar.erl]: ✅

**Comment by ldmberman** [comment]
Linking one of the old tickets to the PR, https://github.com/ArweaveTeam/arweave/issues/25.

File: src/ar_storage.erl (hunk 5 above)
Comment (martin-torhage) [reply src/ar_storage.erl]: ✅

File: src/ar_storage.erl (hunk 6 above)
Comment (martin-torhage) [reply src/ar_storage.erl]: ✅
